cpp/include/cudf/detail/join/filtered_join.cuh

```diff
@@ -0,0 +1,359 @@
+    cuco::pair{std::numeric_limits<hash_value_type>::max(), lhs_index_type{JoinNoneValue}}};
+  build_properties _build_props;           ///< Properties of the build table
+  cudf::table_view _build;                 ///< input table to build the hash map
+  cudf::null_equality const _nulls_equal;  ///< whether to consider nulls as equal
+  std::shared_ptr<cudf::experimental::row::equality::preprocessed_table>
+    _preprocessed_build;  ///< input table preprocssed for row operators
+
+ public:
+  /**
+   * @brief Calculates the required storage size for the hash table
+   *
+   * Computes the appropriate size for the bucket storage based on the input
+   * table size and desired load factor.
+   *
+   * @param tbl Table for which to calculate storage
+   * @param load_factor Target load factor for the hash table
+   * @return Calculated bucket storage size
+   */
+  auto compute_bucket_storage_size(cudf::table_view tbl, double load_factor);
+
+ public:
+  /**
+   * @brief Constructor for filtered_join base class
+   *
+   * Initializes the hash table with the build table and prepares it for join operations.
+   *
+   * @param build The table to build the hash table from
+   * @param compare_nulls How null values should be compared
+   * @param load_factor Target load factor for the hash table
+   * @param stream CUDA stream on which to perform operations
+   */
+  filtered_join(cudf::table_view const& build,
+                cudf::null_equality compare_nulls,
+                double load_factor,
+                rmm::cuda_stream_view stream);
+
+  /**
+   * @brief Populates the hash table with the build table
+   *
+   * @tparam CGSize CUDA cooperative group size
+   * @tparam Ref Reference type for the hash table
+   * @param insert_ref Reference to the hash table for insertion
+   * @param stream CUDA stream on which to perform operations
+   */
+  template <int32_t CGSize, typename Ref>
+  void insert_build_table(Ref const& insert_ref, rmm::cuda_stream_view stream);
+
+  /**
+   * Virtual semi join function overridden in derived classes
+   */
+  virtual std::unique_ptr<rmm::device_uvector<cudf::size_type>> semi_join(
+    cudf::table_view const& probe,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) = 0;
+
+  /**
+   * Virtual anti join function overridden in derived classes
+   */
+  virtual std::unique_ptr<rmm::device_uvector<cudf::size_type>> anti_join(
+    cudf::table_view const& probe,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) = 0;
+};
+
+/**
+ * @brief Implementation of filtered join using set hash tables
+ *
+ * This class extends the base filtered_join to implement join operations
+ * using set semantics, where duplicate keys are not allowed in the hash table.
+ * This implementation is more memory efficient when the same filter table (right table)
+ * is to be reused for multiple semi/anti join operations.
+ */
+class filtered_join_with_set : public filtered_join {
```

Comment:
I'm thinking of whether this separate derived class strategy vs having only one template class with the hash table type as template parameter. By using that template class strategy, we can share code across different hash table types and specialize only if needed.